File: backend/test-db-connection.js
const { Pool } = require('pg');
require('dotenv').config();

console.log('🔍 Testing PostgreSQL connection...\n');

// Display current configuration (without password)
console.log('📋 Current configuration:');
console.log(`Host: ${process.env.DB_HOST || 'localhost'}`);
console.log(`Port: ${process.env.DB_PORT || 5432}`);
console.log(`Database: ${process.env.DB_NAME || 'edtech_platform'}`);
console.log(`User: ${process.env.DB_USER || 'postgres'}`);
console.log(`Password: ${process.env.DB_PASSWORD ? '***SET***' : '***NOT SET***'}\n`);

// Database configuration
const dbConfig = {
  host: process.env.DB_HOST || 'localhost',
  port: process.env.DB_PORT || 5432,
  database: 'postgres', // Connect to default postgres database first
  user: process.env.DB_USER || 'postgres',
  password: process.env.DB_PASSWORD || 'your_password',
};

console.log('🔌 Attempting to connect to PostgreSQL...');

const pool = new Pool(dbConfig);

pool.on('error', (err) => {
  console.error('❌ Unexpected error on idle client', err);
  process.exit(-1);
});

async function testConnection() {
  try {
    const client = await pool.connect();
    console.log('✅ Successfully connected to PostgreSQL!');
    
    // Test basic query
    const result = await client.query('SELECT version()');
    console.log('📊 PostgreSQL version:', result.rows[0].version);
    
    // Check if our target database exists
    const dbCheck = await client.query(
      "SELECT 1 FROM pg_database WHERE datname = $1",
      [process.env.DB_NAME || 'edtech_platform']
    );
    
    if (dbCheck.rows.length > 0) {
      console.log(`✅ Database '${process.env.DB_NAME || 'edtech_platform'}' exists`);
    } else {
      console.log(`⚠️  Database '${process.env.DB_NAME || 'edtech_platform'}' does not exist`);
      console.log('💡 Run "npm run db:setup" to create it');
    }
    
    client.release();
    await pool.end();
    
    console.log('\n🎉 Connection test completed successfully!');
    return true;
    
  } catch (error) {
    console.error('\n❌ Connection failed with error:');
    console.error('Error code:', error.code);
    console.error('Error message:', error.message);
    
    if (error.code === '28P01') {
      console.log('\n💡 This is a password authentication error.');
      console.log('Please check:');
      console.log('1. Your DB_PASSWORD in .env file');
      console.log('2. That the password matches your PostgreSQL user password');
      console.log('3. That the DB_USER is correct');
    } else if (error.code === 'ECONNREFUSED') {
      console.log('\n💡 PostgreSQL server is not running or not accessible.');
      console.log('Please start PostgreSQL service.');
    } else if (error.code === 'ENOTFOUND') {
      console.log('\n💡 Cannot find PostgreSQL server.');
      console.log('Please check your DB_HOST setting.');
    }
    
    await pool.end();
    return false;
  }
}

// Run the test
testConnection()
  .then((success) => {
    process.exit(success ? 0 : 1);
  })
  .catch((error) => {
    console.error('❌ Test failed:', error);
    process.exit(1);
  }); 

File: mlservices/test_db_connection.py
# ============================================================================
# DATABASE CONNECTION TEST
# ============================================================================
# Tests database connection before generating training data
# ============================================================================

import psycopg2
import os
from pathlib import Path

def test_database_connection():
    """Test database connection with different configurations"""
    
    # Try different password configurations
    configs = [
        {
            'host': 'localhost',
            'database': 'school_management',
            'user': 'postgres',
            'password': '1234',
            'port': 5432
        },
        {
            'host': 'localhost',
            'database': 'school_management',
            'user': 'postgres',
            'password': 'password',
            'port': 5432
        },
        {
            'host': 'localhost',
            'database': 'school_management',
            'user': 'postgres',
            'password': '',
            'port': 5432
        }
    ]
    
    print("🔍 Testing database connections...")
    print("=" * 50)
    
    for i, config in enumerate(configs, 1):
        print(f"\n📊 Test {i}: Trying password '{config['password']}'")
        
        try:
            conn = psycopg2.connect(**config)
            cursor = conn.cursor()
            
            # Test basic query
            cursor.execute("SELECT version();")
            version = cursor.fetchone()
            
            print(f"✅ Connection successful!")
            print(f"   Database: {config['database']}")
            print(f"   User: {config['user']}")
            print(f"   PostgreSQL version: {version[0]}")
            
            # Test if attendance tables exist
            cursor.execute("""
                SELECT table_name 
                FROM information_schema.tables 
                WHERE table_schema = 'public' 
                AND table_name LIKE 'attendance%'
                ORDER BY table_name;
            """)
            
            tables = cursor.fetchall()
            if tables:
                print(f"   ✅ Attendance tables found: {[t[0] for t in tables]}")
            else:
                print(f"   ⚠️ No attendance tables found")
            
            cursor.close()
            conn.close()
            
            # Return the working config
            return config
            
        except Exception as e:
            print(f"   ❌ Connection failed: {e}")
    
    print("\n❌ All connection attempts failed!")
    return None

def create_database_if_not_exists():
    """Create database if it doesn't exist"""
    try:
        # Connect to default postgres database
        conn = psycopg2.connect(
            host='localhost',
            database='postgres',
            user='postgres',
            password='1234',
            port=5432
        )
        conn.autocommit = True
        cursor = conn.cursor()
        
        # Check if school_management database exists
        cursor.execute("SELECT 1 FROM pg_database WHERE datname = 'school_management'")
        exists = cursor.fetchone()
        
        if not exists:
            print("📊 Creating school_management database...")
            cursor.execute("CREATE DATABASE school_management")
            print("✅ Database created successfully!")
        else:
            print("✅ Database already exists!")
        
        cursor.close()
        conn.close()
        return True
        
    except Exception as e:
        print(f"❌ Error creating database: {e}")
        return False

if __name__ == '__main__':
    print("🚀 Database Connection Test")
    print("=" * 50)
    
    # First try to create database if it doesn't exist
    create_database_if_not_exists()
    
    # Test connection
    working_config = test_database_connection()
    
    if working_config:
        print(f"\n🎉 Database connection successful!")
        print(f"Use this configuration:")
        print(f"   DB_HOST={working_config['host']}")
        print(f"   DB_NAME={working_config['database']}")
        print(f"   DB_USER={working_config['user']}")
        print(f"   DB_PASSWORD={working_config['password']}")
        print(f"   DB_PORT={working_config['port']}")
    else:
        print(f"\n❌ No working database configuration found!")
        print("Please check your PostgreSQL installation and credentials.")
